src/madonna/doxydocs/man/man3/findrect.3
.TH findrect.C 3 "28 Sep 2000" "madonna" \" -*- nroff -*-
.ad l
.nh
.SH NAME
findrect.C \- 
.SH SYNOPSIS
.br
.PP
\fC#include <stream.h>\fR
.br
\fC#include "sealib.h"\fR
.br
\fC#include "genpart.h"\fR
.br
\fC#include "part.h"\fR
.br
\fC#include "cost.h"\fR
.br
\fC#include "math.h"\fR
.br
\fC#include "phil_glob.h"\fR
.br
.SS Defines

.in +1c
.ti -1c
.RI "#define \fBBBX_MEAN_X\fR  0"
.br
.ti -1c
.RI "#define \fBBBX_MEAN_Y\fR  1"
.br
.ti -1c
.RI "#define \fBBBX_MAX_X\fR  2"
.br
.ti -1c
.RI "#define \fBBBX_MAX_Y\fR  3"
.br
.ti -1c
.RI "#define \fBBBX_N_CELL\fR  4"
.br
.ti -1c
.RI "#define \fBBBX_AREA\fR  5"
.br
.ti -1c
.RI "#define \fBBBX_AVAIL\fR  6"
.br
.ti -1c
.RI "#define \fBBBX_ARRAY_SIZE\fR  7"
.br
.ti -1c
.RI "#define \fBNOTINITIALIZED\fR  -1"
.br
.ti -1c
.RI "#define \fBMAX\fR(a,b)   ((a)>(b)?(a):(b))"
.br
.ti -1c
.RI "#define \fBMEAN_CELLS_PER_AREA\fR  3.0"
.br
.ti -1c
.RI "#define \fBMEAN_CELLS_PER_AREA_IF_CHANNELED\fR  10.0"
.br
.ti -1c
.RI "#define \fBMAXIMUM_NUMBER_OF_PLACEMENT_AREAS\fR  512"
.br
.ti -1c
.RI "#define \fBTOO_SMALL\fR  0.00001"
.br
.ti -1c
.RI "#define \fBTOO_BIG\fR  10000.0"
.br
.ti -1c
.RI "#define \fBVERY_NEGATIVE\fR  -10000"
.br
.in -1c
.SS Enumerations

.in +1c
.ti -1c
.RI "enum \fBRoundingPolicy\fR { \fBNoRounding\fR, \fBRoundUp\fR, \fBRoundDown\fR, \fBRoundNear\fR }"
.br
.in -1c
.SS Functions

.in +1c
.ti -1c
.RI "int \fBfindrectangleandcostfunctions\fR (CIRCUITPTR circuit, int *nx, int *ny, \fBCFUNC\fR *costfunctions, int calldepth)"
.br
.in -1c
.SS Variables

.in +1c
.ti -1c
.RI "int \fBmakeChannels\fR = 0"
.br
.ti -1c
.RI "int \fBrequestedGridPoints\fR [] = {NOTINITIALIZED, NOTINITIALIZED, NOTINITIALIZED}"
.br
.ti -1c
.RI "int \fBexpandableDirection\fR = NOTINITIALIZED"
.br
.ti -1c
.RI "double \fBextraplaza\fR = 1.0"
.br
.in -1c
.SH DEFINE DOCUMENTATION
.PP 
.SS #define BBX_AREA  5
.PP
Definition at line 19 of file findrect.C.
.SS #define BBX_ARRAY_SIZE  7
.PP
Definition at line 21 of file findrect.C.
.SS #define BBX_AVAIL  6
.PP
Definition at line 20 of file findrect.C.
.SS #define BBX_MAX_X  2
.PP
Definition at line 16 of file findrect.C.
.SS #define BBX_MAX_Y  3
.PP
Definition at line 17 of file findrect.C.
.SS #define BBX_MEAN_X  0
.PP
Definition at line 14 of file findrect.C.
.SS #define BBX_MEAN_Y  1
.PP
Definition at line 15 of file findrect.C.
.SS #define BBX_N_CELL  4
.PP
Definition at line 18 of file findrect.C.
.SS #define MAX(a, b)  ((a)>(b)?(a):(b))
.PP
Definition at line 43 of file findrect.C.
.SS #define MAXIMUM_NUMBER_OF_PLACEMENT_AREAS  512
.PP
Definition at line 48 of file findrect.C.
.SS #define MEAN_CELLS_PER_AREA  3.0
.PP
Definition at line 46 of file findrect.C.
.SS #define MEAN_CELLS_PER_AREA_IF_CHANNELED  10.0
.PP
Definition at line 47 of file findrect.C.
.SS #define NOTINITIALIZED  -1
.PP
Definition at line 23 of file findrect.C.
.SS #define TOO_BIG  10000.0
.PP
Definition at line 245 of file findrect.C.
.SS #define TOO_SMALL  0.00001
.PP
Definition at line 244 of file findrect.C.
.SS #define VERY_NEGATIVE  -10000
.PP
Definition at line 375 of file findrect.C.
.SH ENUMERATION TYPE DOCUMENTATION
.PP 
.SS enum RoundingPolicy
.PP
\fBEnumeration values:\fR
.in +1c
.TP
\fB\fINoRounding\fR \fR
.TP
\fB\fIRoundUp\fR \fR
.TP
\fB\fIRoundDown\fR \fR
.TP
\fB\fIRoundNear\fR \fR
.PP
Definition at line 25 of file findrect.C.
.SH FUNCTION DOCUMENTATION
.PP 
.SS int findrectangleandcostfunctions (CIRCUITPTR circuit, int * nx, int * ny, \fBCFUNC\fR * costfunctions, int calldepth)
.PP
Definition at line 74 of file findrect.C.
.SH VARIABLE DOCUMENTATION
.PP 
.SS int expandableDirection = NOTINITIALIZED
.PP
Definition at line 60 of file findrect.C.
.SS double extraplaza = 1.0
.PP
Definition at line 61 of file findrect.C.
.SS int makeChannels = 0
.PP
Definition at line 40 of file findrect.C.
.SS int requestedGridPoints[] = {NOTINITIALIZED, NOTINITIALIZED, NOTINITIALIZED}
.PP
Definition at line 59 of file findrect.C.
.SH AUTHOR
.PP 
Generated automatically by Doxygen for madonna from the source code.

src/RELEASENOTES

This file is maintained by: Arjan van Genderen
                            e-mail: arjan@cas.et.tudelft.nl

Sep. 22, 1999
-------------

Features compared to earlier distributions:
  
   - Seadali now also runs when it can allocate only a 
     reduced set of colors. (In the past, this caused
     that seadali could not be used at all on some
     screens.)
   - The maximum length of cell, instance, terminal 
     and net names has been increased from 14 to 256.
   - Net names are now used by trout to create labels
     in the layout.  As a result, the extracted circuit
     will have the same net (node) names as the original
     circuit.
   - Introduced an installation script 'ocean_install'.

Source code of this new OCEAN distribution is currently
not available.   The old OCEAN distribution is available 
from ftp://cas.et.tudelft.nl/pub/ocean.old.


Dec. 3, 1999
------------

Source code for compiling the cacd and OCEAN trees is now 
also available.  See the file COMPILATION.

Installation script was renamed oceanInstall.


Dec. 17, 1999
-------------

When the "no routing" option was used with trout, labels
were deleted from the layout.  This has now been fixed.


Jan. 5, 2000
------------

Another bug fix for handling labels during routing.


April 14, 2000
--------------

Fixed the problem that kissis didn't work when 
$SIS/sis_lib/.[sm]isrc are not available.


April 19, 2000
--------------

Fixed problem with directory names with more than 100 
characters.


June 8, 2000
------------

Fixed core dump for trout that appeared under Linux.


Sep. 12, 2000
-------------

- Added tools getldm and layflat.
- Added manual pages of layflat, fish, dofunc, mksls


Sep. 27, 2000
-------------

- Made extension in ICELLS for new cacd release that will use xcontrol 
  instead of macro and device


June 12, 2001
-------------

- Compiled with new version of cacd software.


March 13, 2002
--------------

- Compiled with new version of cacd software.
- Bug fix for seadali for deleting terminal in area.
- Let trout place and route master pins that are not connected to
  any instance.


August 12, 2002
---------------

- Updated celllibs/fishbone/src/nelsis3/template/sls_prototypes/oplib.ext



src/madonna/partitioner/findrect.C
/*
 * @(#)findrect.C 1.13 08/20/96 Delft University of Technology
 */


#include <stream.h>
#include "sealib.h"
#include "genpart.h"
#include "part.h"
#include "cost.h"
#include "math.h"
#include "phil_glob.h"

#define BBX_MEAN_X     0
#define BBX_MEAN_Y     1
#define BBX_MAX_X      2
#define BBX_MAX_Y      3
#define BBX_N_CELL     4
#define BBX_AREA       5
#define BBX_AVAIL      6
#define BBX_ARRAY_SIZE 7

#define NOTINITIALIZED -1

enum RoundingPolicy {NoRounding, RoundUp, RoundDown, RoundNear};

static int oldfashionedwayofdoingthings(CIRCUITPTR circuit, int *nx, int *ny,
					CFUNC *costfunctions);
static void computeMeanBbx(CIRCUITPTR circuit, double *meanX, double *meanY);
static int otherDirection(int dir);
static double gridPointsToCoreCells(int gridpoints,int direction,
				    RoundingPolicy rounding = NoRounding);
static int coreCellsToGridPoints(double coreCells, int direction);
static void computeBbxStatistics(CIRCUITPTR circuit, double bbxStatistics[]);
static void thinkAboutGridSize(double stats[],double extraplaza);

extern "C"
int findrectangleandcostfunctions(CIRCUITPTR circuit, int *nx, int *ny,
				  CFUNC *costfunctions, int calldepth);

extern int makeChannels; // TRUE if we must make routing channels

#ifndef MAX
#   define MAX(a,b) ((a)>(b)?(a):(b))
#endif

#define MEAN_CELLS_PER_AREA 3.0
#define MEAN_CELLS_PER_AREA_IF_CHANNELED 10.0
#define MAXIMUM_NUMBER_OF_PLACEMENT_AREAS 512

static CFUNC thecostfunctions[2] = {(CFUNC)netstatecost,
                                    (CFUNC)netstatecost2};

// Following imported parameters specify the requested size and shape of the
// area for the total placement. The area may be expanded only in the
// expandableDirection. In other words,
// requestedGridPoints[expandableDirection] specifies a minimum length, wheras
// requestedGridPoints[otherDirection(expandableDirection)] specifies a maximum
// length.
extern int requestedGridPoints[]; // set by main from argv
extern int expandableDirection;   // set by main from argv
extern double extraplaza;	  // set by main, either by default or from argv

// This results from requestedGridPoints[] and expandableDirection and the
// actual area occupied by the cells to be placed and the shape of these
// cells... It's all computed in findrectangleandcostfunctions()...:
static int    targetGridPoints[2];
static double targetAreas[2];

// This function takes a look at the number and shape of the cells to be
// placed, and tries to think of a reasonable partitioning scheme. The output
// of the function is the number of partitions in horizontal (nx) and vertical
// (ny) direction, as well as an array of costfunctions that can evaluate the
// cost of a netdistribution throughout such a partitioning space.
int findrectangleandcostfunctions(CIRCUITPTR circuit, int *nx, int *ny,
				  CFUNC *costfunctions, int calldepth)
{
   // return oldfashionedwayofdoingthings(circuit,nx,ny,costfunctions);

   // currently we do not support recursive partitioning...:
   if (calldepth > 1) return NIL;

   double stats[BBX_ARRAY_SIZE];
   computeBbxStatistics(circuit,stats);

   thinkAboutGridSize(stats,extraplaza); // maybe modify requestedGridPoints[] ...

   if (stats[BBX_N_CELL] <= 1) return NIL; // no cells...

   // if we also make routing channels, than this is probably a standard-cell
   // like placement. Because flipflops are about 8 times as big as inverters,
   // we want more cells_per_area if the number of flipflops is small.
   double cells_per_area;
   if (makeChannels && stats[BBX_N_CELL] > 100)
      cells_per_area = MEAN_CELLS_PER_AREA_IF_CHANNELED;
   else
      cells_per_area = MEAN_CELLS_PER_AREA;
   int numberOfPlacementAreas
      = 1 + int((stats[BBX_N_CELL] -1) / cells_per_area);
   if (numberOfPlacementAreas < 2) return NIL; // nothing to partition...
   if (numberOfPlacementAreas > MAXIMUM_NUMBER_OF_PLACEMENT_AREAS)
      // we don\'t want too many areas because it\'s too time consuming...
      numberOfPlacementAreas = MAXIMUM_NUMBER_OF_PLACEMENT_AREAS;

   // The shapeCoefficient tells us something about the shape of the cells:
   double shapeCoef = stats[BBX_MEAN_Y] / stats[BBX_MEAN_X];
   // Now think of a reasonable partitioning. We must be aware that
   //   1. targetAreas[HOR] * targetAreas[VER] == numberOfPlacementAreas
   //   2. if cells are very wide we want targetAreas[HOR] << targetAreas[VER]
   //                                               (and the other way around)
   targetAreas[HOR] = sqrt(numberOfPlacementAreas * shapeCoef);
   targetAreas[VER] = sqrt(numberOfPlacementAreas / shapeCoef);

   // initialize the targetGridPoints...:
   targetGridPoints[HOR] = requestedGridPoints[HOR];
   targetGridPoints[VER] = requestedGridPoints[VER];

   // And now we take into account the area that the user specified...:
   // First make sure that the requested area is large enough. If it is already
   // too large, do not make it smaller...
   double requiredSurface =
      stats[BBX_MEAN_X] * stats[BBX_MEAN_Y] * stats[BBX_N_CELL];
   double requestedSurface =
      requestedGridPoints[HOR] * requestedGridPoints[VER];
   if (requiredSurface > requestedSurface)
   {
      // Expand the requested surface in the direction that the user specified
      // as the "minimum" size. Don\'t touch the other side!
      targetGridPoints[expandableDirection] = 
	 int(0.5 + requestedGridPoints[expandableDirection] *
	     (requiredSurface / requestedSurface));
   }
   // At this point we have two potentially conflicting requirements:
   //  1. targetAreas[HOR..VER] is based on the shapes of the cells to be placed
   //  2. targetGridPoints[HOR..VER] is based on the user-supplied sizes
   // Our problem is to find a targetCoreCells[HOR..VER] such that it fits (in
   // both directions) an integer number of times in targetAreas[HOR..VER]. Of
   // course, targetCoreCells must be as close as possible to targetGridPoints
   // AND it must respect the expandableDirection specified by the user...
   //
   //                     ____ D O N \'T   P A N I C ____

   // first create and initialize targetCoreCells...:
   double targetCoreCells[2];
   targetCoreCells[expandableDirection] =
      gridPointsToCoreCells(targetGridPoints[expandableDirection],expandableDirection);
   int fixedDirection = otherDirection(expandableDirection);
   targetCoreCells[fixedDirection] =
      gridPointsToCoreCells(targetGridPoints[fixedDirection],fixedDirection);

   // Well, actually we are only interested in the number of image core cells
   // in a "scaled" fashion.   Suppose most cells have a height of 10 times a
   // core cell, then we rather have something like:
   targetCoreCells[HOR] =
      targetCoreCells[HOR] / MAX(1.0, gridPointsToCoreCells((int)stats[BBX_MEAN_X],HOR));
   targetCoreCells[VER] =
      targetCoreCells[VER] / MAX(1.0, gridPointsToCoreCells((int)stats[BBX_MEAN_Y],VER));

   // Find out what the critical direction is. This is the direction where the
   // "resolution" is the smallest, i.e. where targetCoreCells/targetAreas is
   // smallest...:
   int criticalDirection;
   if (float(targetCoreCells[HOR])/targetAreas[HOR] <
       float(targetCoreCells[VER])/targetAreas[VER])
      criticalDirection = HOR;
   else
      criticalDirection = VER;

   if (targetAreas[criticalDirection] >= targetCoreCells[criticalDirection])
   {
      // we have more areas than we have core cells in the critical direction...
      if (criticalDirection == expandableDirection)
	 // fortunately, we can expand the number of core cells....:
	 targetCoreCells[criticalDirection] = int(targetAreas[criticalDirection]);
      else
      {
	 // not allowed to expand number of core cells.... the only thing we
	 // can do here is lower targetAreas to match targetCoreCells...:
	 targetAreas[criticalDirection] = targetCoreCells[criticalDirection];
	 targetAreas[expandableDirection] =
	    float(numberOfPlacementAreas)/targetAreas[criticalDirection];
      }
   }
   else // I think this "else" part is rubbish...:
   {
      // we have more core cells than areas in the critical direction...
      if (criticalDirection == expandableDirection)
	 // it\'s our lucky day: we are allowed to increase targetCoreCells...:
	 targetCoreCells[criticalDirection] =
	    int(ceil(targetCoreCells[criticalDirection]/targetAreas[criticalDirection])
		* targetAreas[criticalDirection]);
      else
      {
	 // not allowed to expand, so we make it smaller...:
	 double shapeCorrection =
	    floor(targetCoreCells[criticalDirection]/targetAreas[criticalDirection]);
	 targetCoreCells[criticalDirection] =
	    int(targetAreas[criticalDirection] * shapeCorrection);
	 targetCoreCells[expandableDirection] =
	    int(targetAreas[expandableDirection] / shapeCorrection);	    
      }
   }
   // output the results....:
   *nx = MAX(1,int(targetAreas[HOR] + 0.5));
   *ny = MAX(1,int(targetAreas[VER] + 0.5));

   cout << "------ partitioning " << int(stats[BBX_N_CELL])
        << " cells into (hor x vert) = ("
 	<< *nx << " x " << *ny << ") partitions\n" << endl;

   // ...and make sure that requestedGridPoints[] is consistent with our choice
   // of number of areas in horizontal (nx) and vertical (ny) direction:
   int k = 1;
   for (; *nx * coreCellsToGridPoints(k,HOR) < requestedGridPoints[HOR]; ++k)
      ;
   // k is the number of core cells per area in horizontal direction:
   requestedGridPoints[HOR] = (*nx * coreCellsToGridPoints(k,HOR));
   for (k=1; *ny * coreCellsToGridPoints(k,VER) < requestedGridPoints[VER]; ++k)
      ;
   // k is the number of core cells per area in vertical direction:
   requestedGridPoints[VER] = (*ny * coreCellsToGridPoints(k,VER));

   // on the other hand, we loose badly if all the code above somehow arranged
   // for a grid that in one or two dimensions is smaller than the largest cell
   // that we are going to place... Assert that this is not the case:
   if (requestedGridPoints[HOR] < stats[BBX_MAX_X])
      requestedGridPoints[HOR] =
	 coreCellsToGridPoints
	 (gridPointsToCoreCells(int(ceil(stats[BBX_MAX_X])),RoundUp),HOR);

   if (requestedGridPoints[VER] < stats[BBX_MAX_Y])
      requestedGridPoints[VER] =
	 coreCellsToGridPoints
	 (gridPointsToCoreCells(int(ceil(stats[BBX_MAX_Y])),RoundUp),VER);
   cout << "------ requested grid points = (" << requestedGridPoints[HOR]
	<< ", "	<< requestedGridPoints[VER] << ")\n" << endl;

   costfunctions[COSTSTATE] = thecostfunctions[COSTSTATE];
   costfunctions[COSTVEC]   = thecostfunctions[COSTVEC];

   return TRUE;
}


#define TOO_SMALL 0.00001
#define TOO_BIG   10000.0

// This function tries to think of reasonable grid sizes when the user did not
// specify anything...
static void thinkAboutGridSize(double stats[],double extraplaza)
{
   if (extraplaza < TOO_SMALL || extraplaza > TOO_BIG)
      err(5,"unreasonable value for extraplaza");

   if (requestedGridPoints[HOR] == NOTINITIALIZED &&
       requestedGridPoints[VER] == NOTINITIALIZED)
   {
      // no preferred size in both directions... assume square...:
      requestedGridPoints[HOR] = requestedGridPoints[VER]
	 = int(0.5 +
	       sqrt(stats[BBX_AREA] * extraplaza));
   }
   else if (requestedGridPoints[HOR] == NOTINITIALIZED)
   {
      requestedGridPoints[HOR] =
	 int(0.5 +
	     sqrt(stats[BBX_AREA] * extraplaza / requestedGridPoints[VER]));
   }
   else if (requestedGridPoints[VER] == NOTINITIALIZED)
   {
      requestedGridPoints[VER] =
	 int(0.5 + 
	     sqrt(stats[BBX_AREA] * extraplaza / requestedGridPoints[HOR]));
   }

   // check for extreme cases...:
   if (requestedGridPoints[HOR] < stats[BBX_MAX_X])
      requestedGridPoints[HOR] = int(ceil(stats[BBX_MAX_X]));

   if (requestedGridPoints[VER] < stats[BBX_MAX_Y])
      requestedGridPoints[VER] = int(ceil(stats[BBX_MAX_Y]));
      
   // round requestedGridPoints to a multiple of the core cell size...:
   requestedGridPoints[HOR] =
      int(gridPointsToCoreCells(requestedGridPoints[HOR],HOR,RoundUp));
   requestedGridPoints[HOR] = MAX(1,coreCellsToGridPoints(requestedGridPoints[HOR],HOR));
   requestedGridPoints[VER] =
      int(gridPointsToCoreCells(requestedGridPoints[VER],VER,RoundUp));
   requestedGridPoints[VER] = MAX(1,coreCellsToGridPoints(requestedGridPoints[VER],VER));

   // If for some reason the requested area is far too small, increase it to
   // match a magnification of newAreaQuotient ...:
   const double tooSmallAreaQuotient = 0.5;
   const double newAreaQuotient = 0.7;
   const double areaQuotient =
      (requestedGridPoints[HOR] * requestedGridPoints[VER]) /
	 (stats[BBX_N_CELL] * stats[BBX_MEAN_X] * stats[BBX_MEAN_Y]);
   if (areaQuotient < tooSmallAreaQuotient)
   {
      requestedGridPoints[expandableDirection] =
	 int(double(requestedGridPoints[expandableDirection]) *
	     (newAreaQuotient / areaQuotient));
   }
}


// this static variable contains the sizes of a core cell...:
static int coreCellSize[] = {NOTINITIALIZED,NOTINITIALIZED,NOTINITIALIZED};

static int coreCellOverlap[] = {NOTINITIALIZED,NOTINITIALIZED,NOTINITIALIZED};

// convert #gridpoints to #core cells...:
static double gridPointsToCoreCells(int gridpoints,int direction,
				    RoundingPolicy round)
{
   if (gridpoints <= 0) return 0;

   if (coreCellSize[HOR] == NOTINITIALIZED)
   {
      IMAGEDESC *id = getImageDesc(); // get stuff from image.seadif
      coreCellSize[HOR]    = id->size[HOR];
      coreCellSize[VER]    = id->size[VER];
      coreCellOverlap[HOR] = id->overlap[HOR];
      coreCellOverlap[VER] = id->overlap[VER];
   }

   double fgridpoints = double(gridpoints - coreCellOverlap[direction]);
   if (fgridpoints < 1.0) fgridpoints = 1.0;
   double fcorecellsize = double(coreCellSize[direction]);
   if (fcorecellsize < 1.0) fcorecellsize = 1.0;

   switch (round)
   {
   case NoRounding:
      return fgridpoints / fcorecellsize;
   case RoundDown:
      return floor(fgridpoints / fcorecellsize);
   case RoundUp:
      return ceil(fgridpoints / fcorecellsize);
   case RoundNear:
      return floor(0.5 + fgridpoints / fcorecellsize);
   default:
      err(5,"(INTERNAL) gridPointsToCoreCells: "
	  "illegal rounding policy specified");
      return -1;		// not reached
   }
}


static int coreCellsToGridPoints(double coreCells, int direction)
{
   if (coreCells <= 0) return 0;

   if (coreCellSize[HOR] == NOTINITIALIZED)
   {
      IMAGEDESC *id = getImageDesc(); // get stuff from image.seadif
      coreCellSize[HOR] = id->size[HOR];
      coreCellSize[VER] = id->size[VER];
   }

   return int(0.5 + coreCellSize[direction] * coreCells);
}



// This function turns the direction 90 degrees...:
static int otherDirection(int dir)
{
   if (dir == HOR) return VER;
   if (dir == VER) return HOR;
   err(5,"(internal error) illegal direction: HOR or VER expected");
   return NIL;
}


#define VERY_NEGATIVE -10000

// Compute some stats of the bounding boxes...:
static void computeBbxStatistics(CIRCUITPTR circuit, double bbxStatistics[])
{
   int totalx=0,totaly=0,totalarea=0,maximumx=VERY_NEGATIVE,maximumy=VERY_NEGATIVE;
   CIRINSTPTR cirinst = circuit->cirinst;

   static STRING tmpcell = cs("Tmp_Cell_");
   // if (tmpcell==NIL) tmpcell = cs("Tmp_Cell_");	// skip these layouts

   // integer n counts the number of circuits, integer n2 counts the total
   // number of layouts that are available for these circuits.
   int n, n2;
   for (n=0, n2=0; cirinst!=NIL; cirinst=cirinst->next, ++n)
   {
      LAYOUTPTR lay = cirinst->circuit->layout;
      // OK, what to do? There can be more than one layout for this cicuit.
      // Maybe it\'s best to take them all into account...:
      int old_n2;
      for (old_n2 = n2; lay!=NIL; lay=lay->next)
      {
	 if (lay->name == tmpcell) continue;
	 n2 += 1;
	 totalx += lay->bbx[HOR];
	 totaly += lay->bbx[VER];
	 totalarea += lay->bbx[HOR] * lay->bbx[VER];
	 if (lay->bbx[HOR] > maximumx) maximumx = lay->bbx[HOR];
	 if (lay->bbx[VER] > maximumy) maximumy = lay->bbx[VER];
      }
      if (old_n2 == n2)		// this should never happen
      {
	 char s[200];
	 sprintf(s,"(INTERNAL) findrect: circuit (%s(%s(%s))) w/o layout",
		 cirinst->circuit->name,cirinst->circuit->function->name,
		 cirinst->circuit->function->library->name);
	 err(5,s);
      }
   }
   // avoid division by zero, even if n == 0 ...:
   bbxStatistics[BBX_MEAN_X] = double(totalx) / MAX(1,n2);
   bbxStatistics[BBX_MEAN_Y] = double(totaly) / MAX(1,n2);
   bbxStatistics[BBX_MAX_X]  = (double) MAX(0,maximumx);
   bbxStatistics[BBX_MAX_Y]  = (double) MAX(0,maximumy);
   bbxStatistics[BBX_AREA]   = (double) totalarea;
   bbxStatistics[BBX_AVAIL]  = double(n) / double(n2);
   bbxStatistics[BBX_N_CELL] = (double) n;
}
